File: docs-examples/example-java/src/test/java/io/micronaut/jms/docs/exceptions/ErrorHandlingSpec.java
package io.micronaut.jms.docs.exceptions;

import io.micronaut.jms.docs.AbstractJmsSpec;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class ErrorHandlingSpec extends AbstractJmsSpec {

    @Test
    void testCustomErrorHandlersAtTheMethodAndClassLevel() {
        ErrorHandlingProducer producer = applicationContext.getBean(ErrorHandlingProducer.class);
        ErrorThrowingConsumer consumer = applicationContext.getBean(ErrorThrowingConsumer.class);
        CountingErrorHandler errorHandler = applicationContext.getBean(CountingErrorHandler.class);
        AccumulatingErrorHandler classLevelErrorHandler = applicationContext.getBean(AccumulatingErrorHandler.class);

        producer.send("throw an error");

        await().atMost(5, SECONDS).until(() -> consumer.messages.size() == 0 &&
                errorHandler.count.get() == 1 &&
                classLevelErrorHandler.exceptions.size() == 1);
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/jms/docs/exceptions/ErrorHandlingSpec.kt
package io.micronaut.jms.docs.exceptions

import io.micronaut.jms.docs.AbstractJmsSpec
import org.junit.jupiter.api.Test

import org.awaitility.Awaitility.await
import java.util.concurrent.TimeUnit

class ErrorHandlingSpec : AbstractJmsSpec() {

    @Test
    fun testCustomErrorHandlersAtTheMethodAndClassLevel() {
        val producer = applicationContext.getBean(ErrorHandlingProducer::class.java)
        val consumer = applicationContext.getBean(ErrorThrowingConsumer::class.java)
        val errorHandler = applicationContext.getBean(CountingErrorHandler::class.java)
        val classLevelErrorHandler = applicationContext.getBean(AccumulatingErrorHandler::class.java)

        producer.push("throw an error")

        await().atMost(5, TimeUnit.SECONDS).until {
            consumer.processed.size == 0 &&
                errorHandler.count.get() == 1 &&
                classLevelErrorHandler.exceptions.size == 1
        }
    }
}
